Docs for creating a loading script for image datasets

stevhliu opened the pull request:
This PR is a first draft of creating a loading script for image datasets. Feel free to let me know if there are any specificities I'm missing for this. 🙂 

To do:
- [x] Document how to create different configurations.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_4783). All of your documentation changes will be reflected on that endpoint.

lhoestq commented on docs/source/image_dataset_script.mdx:
```diff
@@ -0,0 +1,245 @@
+
+Now that you've added some information about your dataset, the next step is to download the dataset and generate the splits.
+
+1. Use the [`DownloadManager.download`] method to download the dataset and any other metadata you'd like to associate with it. This function accepts a URL or a list or dictionary of URLs. In the Food-101 loading script, you'll notice again the URLs are defined earlier in the script.
```
Maybe here we can encourage users to upload their images to the Hub, instead of hosting them somewhere and pass the URL
```suggestion
1. Use the [`DownloadManager.download`] method to download the dataset and any other metadata you'd like to associate with it. This function accepts a name to a file inside the dataset repository (e.g. in the `data/` folder), or a URL to a file hosted elsewhere. It also supports a list or dictionary of file names or URLs. In the Food-101 loading script, you'll notice again the URLs are defined earlier in the script.
```

lhoestq commented on docs/source/image_dataset_script.mdx:
````diff
@@ -0,0 +1,245 @@
+```
+my_dataset/
+├── README.md
+└── my_dataset.py
````
```suggestion
├── my_dataset.py
└── data/  # optional, may contain your images or TAR archives
```

lhoestq commented on docs/source/image_dataset_script.mdx:
```diff
@@ -0,0 +1,245 @@
+1. Use the [`DownloadManager.download`] method to download the dataset and any other metadata you'd like to associate with it. This function accepts a URL or a list or dictionary of URLs. In the Food-101 loading script, you'll notice again the URLs are defined earlier in the script.
+2. After you've downloaded the dataset, use the [`SplitGenerator`] to organize the images and labels in each split. Name each split with a standard name like: `Split.TRAIN`, `Split.TEST`, and `SPLIT.Validation`. 
+
+    In the `gen_kwargs` parameter, use [`DownloadManager.iter_archive`] to specify the `images` file paths to iterate over and load. You can also specify the associated labels in the `metadata_path`. The `images` and `metadata_path` are actually passed onto the next step where you'll actually generate the dataset.
```
```suggestion
    In the `gen_kwargs` parameter, specify the file paths to the `images` to iterate over and load. If necessary, you can use [`DownloadManager.iter_archive`] to iterate over the images in TAR archives. You can also specify the associated labels in the `metadata_path`. The `images` and `metadata_path` are actually passed onto the next step where you'll actually generate the dataset.
```

lhoestq commented on docs/source/image_dataset_script.mdx:
````diff
@@ -0,0 +1,245 @@
+...                 }
+```
+
+## Generate the dataset metadata
````
```suggestion
## Generate the dataset metadata (Optional)
```

lhoestq reviewed (commented):
Awesome thanks ! I added a few comments to explain that users can host their data on the hub along with the dataset script

lhoestq commented on docs/source/image_dataset_script.mdx:
````diff
@@ -0,0 +1,253 @@
+2. For each configuration, provide a name, description, and where to download the images and labels from.
+
+```py
+>>> class Food101(datasets.GeneratorBasedBuilder):
+...     """Food-101 Images dataset"""
+... 
+...     BUILDER_CONFIGS = [
+...         Food101Config(
````
It's maybe better to not use the `>>>`/`...` syntax here, to make it easy for people to copy paste into a .py file ?

mariosasko replied:
This syntax should be ignored if you use the copy-to-clipboard button.

stevhliu replied:
The syntax is also ignored if you try to copy and paste it normally. But I removed it for code snippets that are being copied into a `py` file since I don't think it needs to be doctested :)

lhoestq reviewed (approved):
Thank you ! :)

mariosasko commented:
IMO it would make more sense to add a "Create image dataset" page with two main sections - a no-code approach with `imagefolder` + metadata (preferred way), and with a loading script (advanced). It should be clear when to choose which. If we leave this as-is, the user who jumps straight to the Vision section could be under the impression that writing a loading script is the preferred way to share a vision dataset due to how this subsection starts:
```
Write a dataset loading script to share a dataset.
```
  
Also, I think a note explaining how to make a dataset gated/disable the viewer to hide the data would be beneficial (it's pretty common to require submitting a form to access a CV dataset).

stevhliu commented:
Great suggestion @mariosasko! I added your suggestions, let me know what you think. For gated dataset access, I just added a tip referring users to the relevant docs since it's more of a Hub feature than `datasets` feature.

mariosasko commented:
Thanks, looks much better now :). I would also move the sections explaining how to create an `imagefolder` for the specific task from the [loading page](https://raw.githubusercontent.com/huggingface/datasets/main/docs/source/image_load.mdx) to this one. IMO it makes more sense to have the basic info (imagefolder structure + `load_dataset` call) there + a link to this page for info on how to create an image folder dataset.

stevhliu commented:
Good idea! Moved everything about `imagefolder` + metadata to the create an image dataset section since the `load_dataset` call is the same for different computer vision tasks.

mariosasko reviewed (commented):
Two nits:
* The `image_load` page has no mentions of ImageFolder with metadata now. We should still mention this feature, considering this is the first page in the vision category.
* Let's explain how to share an ImageFolder dataset on the Hub  (embed the folder inside a cloned repo + git workflow)

stevhliu commented:
Thanks for all the feedbacks! 🥰

What do you think about creating how to share an `ImageFolder` dataset in a separate PR? I think we should create a new section under `Vision` for how to share an image dataset.

mariosasko commented on docs/source/image_load.mdx:
````diff
@@ -78,69 +78,18 @@ Load remote datasets from their URLs with the `data_files` parameter:
 ```
 
-`ImageFolder` creates a `label` column, and the label name is based on the directory name. To ignore the `label` column, set `drop_labels=False` as defined in [`~datasets.packaged_modules.imagefolder.ImageFolderConfig`].
+Some datasets have a metadata file (`metadata.jsonl`) associated with it, containing other information about the data like bounding boxes, text captions, and labels. The metadata is automatically loaded when you call [`load_dataset`] and specify `imagefolder`. 
````
We also support metadata files in CSV now.

```suggestion
Some datasets have a metadata file (`metadata.csv`/`metadata.jsonl`) associated with it, containing other information about the data like bounding boxes, text captions, and labels. The metadata is automatically loaded when you call [`load_dataset`] and specify `imagefolder`. 
```

mariosasko reviewed (commented):
I'm OK with creating the How to share an ImageFolder dataset page in a separate PR.
One nit.

lhoestq commented:
I love it thanks ! I think moving forward we can use CSV instead of JSON Lines in the docs ;)

lhoestq commented on docs/source/image_dataset_script.mdx:
````diff
@@ -0,0 +1,347 @@
+```py
+>>> from datasets import load_dataset
+
+>>> dataset = load_dataset("imagefolder", data_dir="/path/to/folder")
````
or `load_dataset("/path/to/folder")`
